armory/adapt/pytorch.py

```diff
@@ -0,0 +1,1126 @@
+        position.append(np.random.randint(i - j + 1))
+    return tuple(position)
+
+
+class PGD_RandomPatch(PGD_Patch2):
+    """
+    Patch is randomly placed in image
+    """
+
+    def gradient(self):
+        self.x = self.x_orig.clone()
+        self.x_background = self.x[self.index]
+        self.x[self.index] = self.delta
+        super().gradient()
+        self.x[self.index] = self.x_background
+
+    def __call__(
+        self,
+        x_orig: torch.Tensor,
+        y_true=None,
+        y_target=None,
+        x_init=None,
+        iters=10,
+    ):
+        # TODO: verify x_orig in domain
+        iters = int(iters)
+        if iters < 0:
+            raise ValueError("iters must be nonnegative")
+        self.call_init(
+            early_stop=False,
+            best_x=None,
+            x_orig=x_orig,
+            y_true=y_true,
+            y_target=y_target,
+            x_init=x_init,
+            iters=iters,
+        )
+
+        if self.y_target is None:  # untargeted
+            self.targeted = False
+            if self.y_true is None:
+                self.y_target = self.model(self.x)
+            else:
+                self.y_target = self.y_true
+        else:
+            self.targeted = True
+
+        # TODO: improve placement of mask
+        #    For now, just use upper left
+        self.position = (0, 0, 0)
+        self.index = (slice(None),) + tuple(
+            slice(i, i + j) for (i, j) in zip(self.position, self.mask_size)
+        )
+        self.delta_orig = self.x_orig[self.index]
+        self.x_zero = self.x_orig.clone()
+        self.x_zero[self.index] = 0
+        self.delta = self.delta_orig.clone()
+        # check that x_orig is in domain
+        if x_init == "random":
+            self.delta = torch.rand(self.delta)
+        elif x_init is not None:
+            raise ValueError(f"x_init {x_init} must be 'random' or None")
+
+        self.optimizer = torch.optim.SGD(
+            [self.delta], lr=self.eps_step
+        )  # should be part of constructor
+
+        # modify delta (patch) instead of x directly
+        self.delta.requires_grad = True
+        for i in range(iters + 1):
+            self.i = i
+            self.position = random_position(self.x_orig, self.mask_size)
+            self.index = (slice(None),) + tuple(
+                slice(i, i + j) for (i, j) in zip(self.position, self.mask_size)
+            )
+            self.optimizer.zero_grad()
+            self.gradient()
+            self.status()  # check for early stop, best value, bad values, etc.
+            if self.early_stop or self.i >= self.iters:
+                break
+
+            self.update()
+
+        # TODO: work on random placement of patch
+
+        return self.best_x
+
+
+# Create graph of input/output variables and connections?
+
+# attack = Attack([
+#     Input(x_orig),
+#     X_init(x_orig,
+#     Output(x_adv),
+# ])(epsilon)
+
+
+class EoTPGD_Linf(PGD_Linf):
```

Comment:
it seems like the number of distinct classes needed is multiplying rapidly. Would you also need a class for EoT L2, EoT Patch, EoT random patch, etc.?

Reply:
I should have added a note on this one: this was meant to demonstrate how to extend a class to use EoT. I would remove this class, actually. Or, if EoT can be added as a wrapper around an existing class, then that would prevent that sort of class explosion.

I need a better way of handling / representing random components - that would enable specification of what you're performing the expectation over.